src/storage/src/render/upsert/types.rs

```diff
@@ -105,18 +95,26 @@ pub fn upsert_bincode_opts() -> BincodeOpts {
 // the updated value we want to write back to the `UpsertStateBackend`
 // implementation.
 #[derive(Debug, Default, Clone)]
-pub struct UpsertValueAndSize {
+pub struct UpsertValueAndSize<O> {
     /// The value, if there was one.
-    pub value: Option<StateValue>,
+    pub value: Option<StateValue<O>>,
     /// The size of original`value` as persisted,
     /// Useful for users keeping track of statistics.
-    pub size: Option<u64>,
+    pub metadata: Option<ValueMetadata<u64>>,
+}
+
+/// Metadata about an existing value in the upsert state backend. Passed
+/// back into `multi_put`.
+#[derive(Copy, Clone, Debug)]
+pub struct ValueMetadata<S> {
+    pub size: S,
+    pub tombstone: bool,
```

Comment:
nit: would be nice to have this as `is_tombstone`, when reading the code had to come back here to check the type.